### ISSUE TITLE
KCP problems - master issue

### ISSUE THREAD

**Issue by calmh** [syncthing/syncthing#4459]
Filing the obvious issue as this will come up once it starts being used. Using KCP on my (W)LAN at home I get about 400 KiB/s from my laptop to the server. Using TCP I get about 50 MiB/s. There's all kinds of reasons for this, chief among them probably the fact that there's a lot of knobs to tune in KCP and we haven't gotten them right yet.

**Comment by AudriusButkevicius** [comment]
I'd tweak the knobs prior to RC. It might also be the pfilter wrapper oppose to knobs.
Can you try tweaking the knobs to see if you can gain some immediate improvements?

**Comment by calmh** [comment]
Frankly it'd be neat if there was someone more than me doing testing on this... For example someone who knows what the knobs mean...

**Comment by AudriusButkevicius** [comment]
There are only a few of them, try 10x'ing one by one and see if it makes a difference.

**Comment by calmh** [comment]
> I'd tweak the knobs prior to RC.

Then please do so. I seriously don't have either the time nor the inclination at the moment.

**Comment by AudriusButkevicius** [comment]
Well, nor do I.

**Comment by DrSchnagels** [comment]
Is there any "knob" to fully disable (forced) at the moment?

**Comment by alyandon** [comment]
Something to disable KCP for the time being would be nice.  It seems that after my  a mixture of Windows and Linux clients on the same network updated they can no longer connect to each other since they seem to prefer KCP over TCP and the KCP connection seems to never complete.

**Comment by AudriusButkevicius** [comment]
It should not have an effect on TCP connecting. Do you have logs, or if you are running linux, can you SIGQUIT the process and capture the stacktraces?

**Comment by DrSchnagels** [comment]
Furthermore I noticed this in syncthing-android and I think it is concerning me that there are servers connected I didn't choose to want to.
On all my devices I use fixed IP addresses and i have disabled any NAT or other discovery functions.
Do you know what this in the log means?
[VJLWT] 13:06:50 INFO: kcp://0.0.0.0:22020 resolved external address kcp://91.50.7.148:22020 (via stun.schlund.de:3478)
Does the KCP protocol use another, different discovery method I have to disable?

Thanks!

**Comment by AudriusButkevicius** [comment]
Yes, it uses STUN to traverse NATs. I think we should tie usage of KCP to NAT traversal setting.

**Comment by alyandon** [comment]
@AudriusButkevicius I can see no evidence that any connection attempt is being made from the logs.  I'll see if I can capture a stacktrace for you.  Would you like me to open a separate issue?

**Comment by AudriusButkevicius** [comment]
Probably, if you can reproduce. You could also run wit STTRACE=connections on both sides, that will have mire meaningful logs.

**Comment by AudriusButkevicius** [comment]
Actually, disabling Stun is already possible by setting StunKeepaliveS to 0.
If you want to disable KCP all together, you can replace listen address from `default` to `tcp://:22000, dynamic+https://relays.syncthing.net/endpoint`

**Comment by alyandon** [comment]
I changed the listen address setting and restarted syncthing and suddenly all my devices were discovering each other and connecting again.

**Comment by DrSchnagels** [comment]
Thanks for the advice AudriusButkevicius.
I thought typing "none" into the listen address field would be enough.

**Comment by AudriusButkevicius** [comment]
@alyandon you should still provide logs for debugging.

**Comment by alyandon** [comment]
@AudriusButkevicius Unfortunately, I don't really have time to troubleshoot this issue further right now.

If I get some free time I'll revert the listen configuration on a couple of nodes and collect the logs and open an issue.  Since I'm reluctant to post logs with private IP address information, node names, etc in a public tracker do you mind if I email them to you instead?

**Comment by AudriusButkevicius** [comment]
Sure, yet if we don't get any help it might be hard for us to fix it, as the limited testing we did, didn't show issues.

The thing I'd recommend checking is the addresses syncthing discovers for the other device.
Looking at discovery stats for the recent time, it seems that some devices announce quite a few kcp addresses, meaning that STUN is probably causing more harm than good.

**Comment by alyandon** [comment]
Sure, I understand you guys need help with troubleshooting this and I'm willing to do the work.  The fact that you don't have dozens of people descending on this thread or opening issues kind of leads me to believe that most people aren't having issues like I was with the update.

And the things about discovery - as far as I could see (even with STTRACE=connections) discovery and tcp connections weren't taking place like they should have been - even for devices in the same layer 2 broadcast domain, even when I forced peers from dynamic setting to tcp://xxx.xxx.xxx.xxx:22000, etc.  I let several client instances run for 30 minutes and it didn't log anything after the initial flurry of activity from startup.  SIGQUIT on one of the Linux instances didn't really reveal anything interesting but I'll make sure to bundle all that up when I get the chance.

**Comment by AudriusButkevicius** [comment]
Make sure you don't get something like this:
https://github.com/syncthing/syncthing/pull/4452#issuecomment-338512379

**Comment by alyandon** [comment]
@AudriusButkevicius Nope, nothing that egregious.  I've gathered and emailed you the logs - let me know if you didn't get them.

**Comment by AudriusButkevicius** [comment]
I did, I hope to go through them today.

**Comment by AudriusButkevicius** [comment]
There are still problems with address announcements:

```
 devices | kcp_addresses
---------+---------------
       1 |           389
       1 |            81
       1 |            49
       1 |            30
       1 |            27
       1 |            25
       1 |            23
       1 |            13
       1 |            10
       2 |             9
       3 |             8
       5 |             7
       7 |             6
      13 |             5
      25 |             4
      67 |             3
     690 |             2
    2186 |             1
```

**Comment by AudriusButkevicius** [comment]
Yet looking at usage reports there are tons of people using their own builds that don't have the DNS resolution fix.

**Comment by AudriusButkevicius** [comment]
Hmm

```
ur=# select transport, sum(count) as conns from (select r.key as transport, r.value::int as count from reports, jsonb_each_text(transport) r where reportversion = 3 and transport <> '{}'::jsonb order by received) v group by 1;
 transport | conns
-----------+-------
 relay     |   124
 tcp       |   615
(2 rows)
```

**Comment by alyandon** [comment]
rc2 appears to have fixed my connectivity issues (really no surprises there).  However, I'm noticing a distinct lack of ipv6 connectivity (no remote ipv6, no link-local ipv6, etc) between dual-stack hosts.

**Comment by DrSchnagels** [comment]
My use-case is (may change in the future) to use fixed ip addresses and no syncing over the internet.
In version .40 i saw that there is a new, second KCP listener now.
Since i have no experiences with KCP id love to see an option to fully disable it.
Or at least that it does not use something like TCP-NAT.
Thanks 👍 :) !

**Comment by AudriusButkevicius** [comment]
Change the default address to what I suggested

**Comment by calmh** [comment]
@AudriusButkevicius since you co-opted this for general stuff, anything left to do here other than the initial "it's slow"?

**Comment by AudriusButkevicius** [comment]
Potentially easier way to disable, and no announcements to discovery on symmetric nat

**Comment by galileolajara** [comment]
Hi,

I'm a syncthing fan for since 2016 and have experiencing slow lan performance (400kb/s) and I'm glad I knew it's because of this. I suggest by default we disable KCP till this problem is fixed? I was surprised that its too slow, given that I was using syncthing before and its fast (syncs more instantly than resilio) and I guess its because there's no KCP before (TCP only)

I hope other users won't see this as a huge problem like mine (syncing 10GB takes forever)

Thank you guys for developing such an awesome piece of software!

**Comment by AudriusButkevicius** [comment]
Preferred connection method is still TCP, if we are able to establish a TCP connection.
KCP is only used if TCP cannot be established. The connection type is shown somewhere in the device panel (when hovering on one of the elements).

**Comment by galileolajara** [comment]
Ok, but I've tried restarting my syncthing (thru web GUI) on my device and my team mates' (yes we use it on our indie game dev, it's great! 😄) several times but still 400kb/s, sometimes I see it go down to 15kb/s. Tried changing WIFI router (different brand), but still...

I'm lost, but I'm glad I found out this thread. I wanted to disable KCP forcefully.

We're using vanilla syncthing zip (no wrappers or inotify) on windows. Does it have something to do with firewalls?
